File: Python Codes/Bisection.py
#Bisection method
#Written by Yusuf Hani Yusuf 202107475

import pandas as pd
from sympy import *

x = Symbol('x')

f = x + 10**-3*exp(20*x) - 9

a = 0.0
b = 1.0
c = 0.0
Tol = 100
Error = 0.00001
k = int(input("Enter the number of iterations: "))
i = 0
a_col=[]
b_col=[]
c_col=[]
i_col=[]
fa_col=[]
fb_col=[]
fc_col=[]
Tol_col=[]

fa = f.subs({x:a})
fb = f.subs({x:b})
if fa*fb > 0:
    print("There is no root in this range")
    exit()

while i<k and Tol>=Error:
    i_str = str(i)
    i_col.append(i_str.zfill(2))
    a_col.append("{:.5g}".format(float(a)))
    b_col.append("{:.5g}".format(float(b)))
    c = (a+b)/2.0
    c_col.append("{:.5g}".format(float(c)))
    fa = f.subs({x:a})
    fa_col.append("{:.5g}".format(float(fa)))
    fb = f.subs({x:b})
    fb_col.append("{:.5g}".format(float(fb)))
    fc = f.subs({x:c})
    fc_col.append("{:.5g}".format(float(fc)))
    if i==0:
        Tol_col.append('-')
        c_temp = c
    else:
        Tol = abs(c-c_temp)
        Tol_col.append("{:.5g}".format(float(Tol)))
        c_temp = c
    if fa*fc > 0:
        a = c
    elif fb*fc > 0:
        b = c
    i += 1
table = pd.DataFrame({
    'i': i_col,
    'a': a_col,
    'b': b_col,
    'c': c_col,
    'f(a)': fa_col,
    'f(b)': fb_col,
    'f(c)': fc_col,
    'Tol': Tol_col
})
print("")
print("Bisection method:",'\n')
print(table.to_markdown(index=False),'\n')
if k>i:
    print("You have exceeded the maximum number of iterations according to the specified Tolerance!!",'\n')
print("The roots is", "{:.5g}".format(float(c)))

File: Matlab_Codes/Bisection.m
%Bisection Method
%Written by Yusuf Hani Almoadhen 

clc
clear
close all

a=0.3;
b=0.7;
c=0;

f = @(x) x+10^-3*exp(20*x)-9;
index=1;
i=0;
k=input("Enter the number of iterations: ");
Error=10^-4;
tol=100;

K = [];
A = [];
B = [];
C = [];
Fa = [];
Fb = [];
Fc = [];
Tol = [];

while i<k && Error<=tol
    K(index,1) = i;
    A(index,1) = a;
    B(index,1) = b;
    c = (a+b)/2;
    C(index,1) = c;
    fa = f(a);
    Fa(index,1) = fa;
    fb = f(b);
    Fb(index,1) = fb;
    fc = f(c);
    Fc(index,1) = fc;
    if i == 0
        Tol(index,1) = "-";
    else
        tol = abs(c-c_temp);
        Tol(index,1) = tol;
    end
    if fa*fc > 0
        a = c;
    else
        b = c;
    end
    c_temp = c;
    index = index+1;
    i = i+1;
end
fprintf('\n')
T = table(K,A,B,C,Fa,Fb,Fc,Tol);
disp(T)
fprintf('\n')
if k>index
    disp("You have exceeded the number of iterations accroding to the specified Tolerance!!")
    fprintf('\n')
end
disp("The root is " + num2str(c))
